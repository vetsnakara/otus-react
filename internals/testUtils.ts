import { ShallowWrapper, ReactWrapper } from 'enzyme'

type Wrapper = ShallowWrapper | ReactWrapper

export const findByDataAttr = (wrapper: Wrapper, val: string) =>
  wrapper.find(`[data-test="${val}"]`)

export const veryLongTitle = Array.from(Array(20).keys())
  .map(() => 'Very long title.')
  .join(' ')
