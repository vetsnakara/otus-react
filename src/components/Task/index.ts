export { default } from './Task'
export * from './Task'
