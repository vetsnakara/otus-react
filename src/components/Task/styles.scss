.task {
  display: flex;
  align-items: center;

  padding: 5px 15px;
  background-color: #fff;

  &__title {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    margin-right: 10px;

    &.completed {
      text-decoration: line-through;
    }
  }

  &__btn {
    flex-shrink: 0;
    position: relative;
    width: 30px;
    height: 30px;
    border: none;
    border-radius: 50%;
    background-color: rgba(0, 0, 0, 0.05);
    cursor: pointer;
    outline: none;

    & > svg {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
    }

    & path {
      fill: rgba(0, 0, 0, 0.5)
    }

    &--toggle {
      margin-right: 10px;
      transition: all .2s ease;

      &.completed {
        background-color: #4dd599;

        & path {
          fill: #fff
        }
      }

      svg {
        opacity: 0;
        transition: all .2s ease;
      }

      &:hover,
      &.completed {
        & > svg {
          opacity: 1;
        }
      }
    }

    &--remove {
      visibility: hidden;
      margin-left: auto;
      transition: all .2s ease;
      transform: scale(0);
      opacity: 0;

      .task:hover & {
        visibility: visible;
        opacity: 1;
        transform: scale(1);
      }
    }
  }

  &__icon {
    width: 15px;
    height: 15px;
  }
}