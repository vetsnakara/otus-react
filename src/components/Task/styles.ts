import styled from 'styled-components'

import { SVGIcon } from '../SVGIcon'

interface TaskProps {
  completed?: boolean
}

/**
 * TaskContainer
 */
export const TaskContainer = styled.div`
  display: flex;
  align-items: center;

  padding: 5px 15px;
  background-color: #fff;
`

TaskContainer.displayName = 'TaskContainer'

/**
 * Button
 */
const Button = styled.button<TaskProps>`
  flex-shrink: 0;
  position: relative;
  width: 30px;
  height: 30px;
  border: none;
  border-radius: 50%;
  background-color: rgba(0, 0, 0, 0.05);
  cursor: pointer;
  outline: none;

  & > svg {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  & path {
    fill: rgba(0, 0, 0, 0.5);
  }
`
/**
 * ToggleButton
 */
export const ToggleButton = styled(Button)`
  margin-right: 10px;
  transition: all 0.2s ease;

  svg {
    opacity: 0;
    transition: all 0.2s ease;
  }

  &:hover > svg {
    opacity: 1;
  }

  ${({ completed }) =>
    completed &&
    `background-color: #4dd599;

    & svg {
      opacity: 1;

      path {
        fill: #fff;
      }
    }`}
`

ToggleButton.displayName = 'ToggleButton'

/**
 * RemoveButton
 */
export const RemoveButton = styled(Button)`
  visibility: hidden;
  margin-left: auto;
  transition: all 0.2s ease;
  transform: scale(0);
  opacity: 0;

  ${TaskContainer}:hover & {
    visibility: visible;
    opacity: 1;
    transform: scale(1);
  }
`

RemoveButton.displayName = 'RemoveButton'

/**
 * Title
 */
export const Title = styled.p<TaskProps>`
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  margin-right: 10px;
  transition: all 0.2s ease;

  ${({ completed }) =>
    completed &&
    `
      text-decoration: line-through;
      color: rgba(0, 0, 0, 0.4);
    `}
`

Title.displayName = 'Title'

/**
 * Icon
 */
export const Icon = styled(SVGIcon)`
  width: 15px;
  height: 15px;
`
