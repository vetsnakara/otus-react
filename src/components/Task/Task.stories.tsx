import React from 'react'

import { actions } from '@storybook/addon-actions'
import { withKnobs, text, boolean } from '@storybook/addon-knobs'

import { TaskData } from '../../interfaces'
import Task from './Task'

import { veryLongTitle } from '../../../internals/testUtils'

export default {
  title: 'Task',
  component: Task,
  excludeStories: /.*Data$/,
  decorators: [withKnobs, (story) => <div className="container">{story()}</div>]
}

// default task data
const taskData: TaskData = {
  title: 'My awesome task',
  completed: false
}

// task data with knobs
const taskDataWithKnobs = () => ({
  title: text('title', 'My awesome task'),
  completed: boolean('completed', false)
})

// task actions
const taskActions = actions({
  onToggleComplete: 'Toggle task complete',
  onRemove: 'Remove'
})

/**
 * Default task
 */
export const Default: React.FC<{}> = () => (
  <Task task={{ ...taskDataWithKnobs() }} {...taskActions} />
)

/**
 * Task with long title
 */
export const LongTitle: React.FC<{}> = () => (
  <Task task={{ ...taskData, title: veryLongTitle }} {...taskActions} />
)

/**
 * Competed task
 */
export const Completed: React.FC<{}> = () => (
  <Task task={{ ...taskData, completed: true }} {...taskActions} />
)
