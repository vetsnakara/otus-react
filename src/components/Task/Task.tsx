import React, { FC } from 'react'
import cn from 'classnames'

import type { TaskProps } from 'types'

import { SVGIcon as Icon } from '../SVGIcon'

import './styles.scss'

export const Task: FC<TaskProps> = ({
  task: { title, completed = false },
  onToggleComplete,
  onRemove
}) => (
  <div data-test="task" className="task">
    <button
      data-test="task-toggle-btn"
      className={cn('task__btn', 'task__btn--toggle', {
        completed
      })}
      onClick={onToggleComplete}
    >
      <Icon name="check" className="task__icon" />
    </button>

    <p data-test="task-title" className={cn('task__title', { completed })}>
      {title}
    </p>

    <button
      data-test="task-remove-btn"
      className="task__btn task__btn--remove"
      onClick={onRemove}
    >
      <Icon name="cancel" className="task__icon" />
    </button>
  </div>
)
