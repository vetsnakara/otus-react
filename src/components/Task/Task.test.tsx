import React from 'react'
import { shallow } from 'enzyme'

import type { TaskProps, TaskData, TaskActions } from '../../interfaces'
import { Task } from './Task'

import { findByDataAttr } from '../../../internals/testUtils'

interface SetupProps extends Partial<TaskData>, Partial<TaskActions> {}

/**
 * Task wrapper factory function
 */
const setup = ({
  title = 'Hello',
  completed,
  onToggleComplete = jest.fn(),
  onRemove = jest.fn()
}: SetupProps = {}) => {
  const setupProps: TaskProps = {
    task: {
      title,
      completed
    },
    onToggleComplete,
    onRemove
  }

  return shallow(<Task {...setupProps} />)
}

describe(`Task`, () => {
  test(`Renders correctly`, () => {
    const title = 'Hello'
    const wrapper = setup({ title })

    // task is rendered
    const task = findByDataAttr(wrapper, 'task')
    expect(task.length).toBe(1)

    // toggle btn is rendered
    const completeBtn = findByDataAttr(task, 'task-toggle-btn')
    expect(completeBtn.length).toBe(1)

    // remove btn is rendered
    const removeBtn = findByDataAttr(task, 'task-remove-btn')
    expect(removeBtn.length).toBe(1)

    // title value is expected
    const taskTitle = findByDataAttr(task, 'task-title')
    expect(taskTitle.text()).toContain(title)
  })

  test(`Click 'complete' button`, () => {
    const onToggleComplete = jest.fn()
    const wrapper = setup({ onToggleComplete })

    const completeBtn = findByDataAttr(wrapper, 'task-toggle-btn')
    completeBtn.simulate('click')
    expect(onToggleComplete).toHaveBeenCalledTimes(1)
  })

  test(`Click 'remove' button`, () => {
    const onRemove = jest.fn()
    const wrapper = setup({ onRemove })

    const removeBtn = findByDataAttr(wrapper, 'task-remove-btn')
    removeBtn.simulate('click')
    expect(onRemove).toHaveBeenCalledTimes(1)
  })
})
