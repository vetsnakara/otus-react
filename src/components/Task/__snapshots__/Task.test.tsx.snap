// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Renders correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 15px;
  background-color: #fff;
}

.c2 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 30px;
  height: 30px;
  border: none;
  border-radius: 50%;
  background-color: rgba(0,0,0,0.05);
  cursor: pointer;
  outline: none;
  margin-right: 10px;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c2 > svg {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c2 path {
  fill: rgba(0,0,0,0.5);
}

.c2 svg {
  opacity: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c2:hover > svg {
  opacity: 1;
}

.c5 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  width: 30px;
  height: 30px;
  border: none;
  border-radius: 50%;
  background-color: rgba(0,0,0,0.05);
  cursor: pointer;
  outline: none;
  visibility: hidden;
  margin-left: auto;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  opacity: 0;
}

.c5 > svg {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c5 path {
  fill: rgba(0,0,0,0.5);
}

.c0:hover .c5 {
  visibility: visible;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c4 {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  margin-right: 10px;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c3 {
  width: 15px;
  height: 15px;
}

<div
  className="c0 c1"
>
  <button
    className="styles__Button-sc-11nhwp9-1 c2"
    onClick={[MockFunction]}
  >
    <svg
      className="c3"
      height="100%"
      style={Object {}}
      viewBox="0 0 512 512"
      width="100%"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="M504.502,75.496c-9.997-9.998-26.205-9.998-36.204,0L161.594,382.203L43.702,264.311c-9.997-9.998-26.205-9.997-36.204,0 c-9.998,9.997-9.998,26.205,0,36.203l135.994,135.992c9.994,9.997,26.214,9.99,36.204,0L504.502,111.7 C514.5,101.703,514.499,85.494,504.502,75.496z"
        fill="#000"
      />
    </svg>
  </button>
  <p
    className="c4"
  >
    Hello
  </p>
  <button
    className="styles__Button-sc-11nhwp9-1 c5"
    onClick={[MockFunction]}
  >
    <svg
      className="c3"
      height="100%"
      style={Object {}}
      viewBox="0 0 512.001 512.001"
      width="100%"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="M294.111,256.001L504.109,46.003c10.523-10.524,10.523-27.586,0-38.109c-10.524-10.524-27.587-10.524-38.11,0L256,217.892 L46.002,7.894c-10.524-10.524-27.586-10.524-38.109,0s-10.524,27.586,0,38.109l209.998,209.998L7.893,465.999 c-10.524,10.524-10.524,27.586,0,38.109c10.524,10.524,27.586,10.523,38.109,0L256,294.11l209.997,209.998 c10.524,10.524,27.587,10.523,38.11,0c10.523-10.524,10.523-27.586,0-38.109L294.111,256.001z"
      />
    </svg>
  </button>
</div>
`;
